README.md
# lit-weather

Lit powered weather forecast web component, leveraging the OpenWeatherMap API. It's customizable with tailwindcss, lightweight, and shadow dom free. Replaces [paper-weather](https://github.com/ryanburns23/paper-weather)

![example](/demo/assets/og.png)

## Installation

```bash
pnpm i -S lit-weather
```

## Usage

```tsx
import 'lit-weather';

<lit-weather api-key="API_KEY" query="ZIP_CODE_OR_CITY" />;
```

## Getting an API Key

To use the lit-weather component, you will need an API key from OpenWeatherMap. You can get one [here](https://home.openweathermap.org/api_keys). Afrer you request a key, it take a few minutes to be approved.

Free tier limits are:

- 60 calls/minute
- 1,000,000 calls/month

I left a key exposed in the demo but it is not guaranteed to work.I don't know if they enfore these limits as I have not reached them but have definitely made more than 60 calls/minute. 😄

The component calls both the ["current weather"](https://openweathermap.org/current) and ["3-hour forecast 5 days"](https://openweathermap.org/forecast5) endpoints.

## Styles

Shadow root is turned off, styles are accessible just like any other html element. This is done so that you can style the component to your liking and we can use tailwindcss without shipping any css. We provide classes for each element in the component.

If you do not overwrite the styles and want to use the default tailwind classes, you must either import the `dist/tw.css` file or include the `dist/dist/src/LitWeather.js` in your tw configs content array.

```js
/** @type {import('tailwindcss').Config} */
module.exports = {
  darkMode: 'class',
  content: ['node_modules/lit-weather/dist/src/LitWeather.js'],
  theme: {
    extend: {},
  },
  plugins: [],
};
```

## Frameworks

This should work in any framework that supports web components. If you have any issues, please open an issue.

## Roadmap

- Add more providers ([open-meteo](https://open-meteo.com/), [weatherbit](https://www.weatherbit.io/))

## Aknowledgements

- Weather Icons by [Naman Rastogi](https://www.uplabs.com/posts/google-now-weather-icons-freebie)
- Weather data from [OpenWeatherMap](https://openweathermap.org/)


CHANGELOG.md
## 1.0.1 (2023-03-02)

- add package.json main export, remove exports object

## 1.0.0 (2023-03-02)

- Enable strict mode
- Simplify class names and html structure
- Add preview variant
- Showcase preview variant in demo
- Internal props as @state protected properties
- Fix [issues with class fields](https://lit.dev/docs/components/properties/#avoiding-issues-with-class-fields), use declare and define defaults in constructor
- Remove excess `request.update()` calls
- Hide protected properties from docs
- Add eslint

## 1.0.0-beta.4 (2023-03-01)

- Change bg-color on img components to `bg-inherit`

## 1.0.0-beta.3 (2023-03-01)

- Add "./lit-weather" to package.json exports
- Add `.js` extension to import in `./lit-weather.ts`
- Add bg color to img element

## 1.0.0-beta.2 (2023-03-01)

- Add global JSX.IntrinsicElements def for react
- Include custom-elements.json in package.json files array
- Mark optional properties as optional in typescript

## 1.0.0-beta.1 (2023-02-28)

- Add icons to package.json files array
- Add "demo/assets" to package.json files array

## 1.0.0-beta.0 (2023-02-28)

- Typescript

## 1.0.0-alpha.2 (2020-03-11)

- Add number type support for query property
- dispatch data-changed event

## 1.0.0-alpha.1 (2020-03-10)

- Initial commit


test/lit-weather.test.js
import { html, fixture, expect } from '@open-wc/testing';

import '../lit-weather.js';

describe('LitWeather', () => {
  it('has a default title "Hey there" and counter 5', async () => {
    const el = await fixture(html`
      <lit-weather></lit-weather>
    `);

    expect(el.title).to.equal('Hey there');
    expect(el.counter).to.equal(5);
  });

  it('increases the counter on button click', async () => {
    const el = await fixture(html`
      <lit-weather></lit-weather>
    `);
    el.shadowRoot.querySelector('button').click();

    expect(el.counter).to.equal(6);
  });

  it('can override the title via attribute', async () => {
    const el = await fixture(html`
      <lit-weather title="attribute title"></lit-weather>
    `);

    expect(el.title).to.equal('attribute title');
  });

  it('shows initially the text "hey there Nr. 5!" and an "increment" button', async () => {
    const el = await fixture(html`
      <lit-weather></lit-weather>
    `);

    expect(el).shadowDom.to.equalSnapshot();
  });

  it('passes the a11y audit', async () => {
    const el = await fixture(html`
      <lit-weather></lit-weather>
    `);

    await expect(el).shadowDom.to.be.accessible();
  });
});


src/LitWeather.js
import { html, css, LitElement } from 'lit-element';
import { repeat } from 'lit-html/directives/repeat';

// prettier-ignore
const getLocalTime = (time) => time.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true });
// prettier-ignore
const getWeekDay = index => ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'][index];

export class LitWeather extends LitElement {
  static get styles() {
    return css`
      :host {
        --ps-lit-weather-header-text-color: #212121;
        --ps-lit-weather-text-color: #878787;
        --ps-lit-weather-secondary-text-color: #bababa;
        --ps-lit-weather-spacing-xs: 8px;
        --ps-lit-weather-spacing-s: 12px;
        --ps-lit-weather-spacing-m: 20px;
        --ps-lit-weather-spacing-l: 36px;

        display: block;
        /* System fonts */
        /* prettier-ignore */
        font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';
        color: var(--ps-lit-weather-text-color);
        line-height: 1.4;
        font-size: 14px;
      }

      [part='location'] {
        font-size: x-large;
      }

      [part='today'],
      [part='description'] {
        font-size: medium;
      }

      [part='description'] {
        text-transform: capitalize;
        margin-bottom: var(--ps-lit-weather-spacing-m);
      }

      [part='temp-wrapper'] {
        margin-bottom: var(--ps-lit-weather-spacing-m);
      }

      [part='img-temp-wrapper'] {
        flex: 1;
      }

      [part='temp'] {
        color: var(--ps-lit-weather-header-text-color);
        font-size: 64px;
        line-height: 1;
      }

      [part='img'] {
        margin-right: 10px;
        height: 70px;
        width: 70px;
      }

      [part='unit-toggle'] {
        font-size: 16px;
        margin-top: 6px;
        margin-left: 2px;
      }

      [part='extra-info'] {
        display: flex;
        flex-direction: column;
        font-size: medium;
        flex: 1;
        margin-left: var(--ps-lit-weather-spacing-l);
      }

      [part='forcast-wrapper'] {
        overflow: auto;
      }

      [part='forcast-item'] {
        min-width: 70px;
        padding: 10px;
        color: var(--ps-lit-weather-secondary-text-color);
      }

      [part='forcast-item-title'] {
        text-align: center;
        line-height: 15px;
        margin-bottom: var(--ps-lit-weather-spacing-xs);
      }

      [part='forcast-item-high'],
      [part='forcast-item-low'] {
        line-height: 15px;
        font-size: 13px;
      }

      [part='forcast-item-high'] {
        color: var(--ps-lit-weather-text-color);
      }

      [part='error-img'] {
        height: 120px;
        width: 120px;
        margin-bottom: var(--ps-lit-weather-spacing-s);
      }
      [part='error-text'] {
        text-align: center;
        font-size: x-large;
      }

      .mobile-header {
        display: none;
      }

      @keyframes pulse {
        0% {
          transform: scale(1);
          opacity: 0;
        }
        50% {
          transform: scale(1.2);
          opacity: 1;
        }
      }
      [part='loading'] {
        text-align: center;
      }
      [part='loading'] > * {
        animation: pulse 1s infinite ease-in-out;
        will-change: opacity, transform;
        display: inline-block;
        padding: 64px;
      }

      @media (max-width: 390px) {
        [part='img-temp-wrapper'] {
          flex: 1;
        }
        [part='extra-info'] {
          font-size: 12px;
          min-width: 90px;
          max-width: 90px;
        }

        [part='img'] {
          margin-right: 5px;
          height: 48px;
          width: 48px;
        }

        [part='temp'] {
          font-size: 48px;
          letter-spacing: -1px;
        }

        [part='unit-toggle'] {
          font-size: 13px;
          margin-top: 2px;
          margin-left: 2px;
        }

        [part='extra-info'] {
          margin-left: var(--ps-lit-weather-spacing-s);
        }

        [part='location'] {
          color: var(--ps-lit-weather-header-text-color);
          font-size: 18px;
        }

        [part='today'],
        [part='description'] {
          display: none;
        }

        .mobile-header {
          display: block;
          font-size: 12px;
          margin-bottom: 26px;
        }
      }

      /* Helper Classes */
      * {
        box-sizing: border-box;
      }
      .horizontal {
        display: flex;
        flex-direction: row;
      }
      .center {
        align-items: center;
      }
      .center-center {
        align-items: center;
        justify-content: center;
      }
      .capitalize {
        text-transform: capitalize;
      }
    `;
  }

  render() {
    return this.loading ? this.loadingTm : this.pageContent;
  }

  get pageContent() {
    return html`
      ${this.error ? this.errorTm : this.content}
    `;
  }

  get content() {
    return html`
      <div part="location">${this.data.name}</div>
      <div part="today">${this.weekDayName} ${this.time}</div>
      <div part="description">${this.data.description}</div>

      <div class="mobile-header">
        ${this.weekDayName.substring(0, 3)} ${this.time.substring(0, 2)}
        ${this.time.substring(this.time.length - 2, this.time.length)},
        <span class="capitalize">${this.data.description}</span>
      </div>

      <div part="temp-wrapper" class="horizontal center">
        <img
          part="img"
          draggable="false"
          src="${this.data.weatherImgSrc}"
          loading="lazy"
          height="70"
          width="70"
        />

        <div part="img-temp-wrapper" class="horizontal">
          <div part="temp">${this.data.temp}</div>
          <div part="unit-toggle">
            <span @click="${this.toggleUnits}">°F</span>
            <span> | </span>
            <span @click="${this.toggleUnits}">°C</span>
          </div>
        </div>

        <div part="extra-info">
          <div>Sunrise: ${this.data.sunrise}</div>
          <div>Sunset: ${this.data.sunset}</div>
          <div>Humidity: ${this.data.humidity}%</div>
          <div>Wind: ${this.data.wind}</div>
        </div>
      </div>

      <div part="forcast-wrapper" class="horizontal">
        ${repeat(
          this.forecast,
          item => html`
            <div part="forcast-item">
              <div part="forcast-item-title">${item.weekDayName.substring(0, 3)}</div>
              <img
                part="forcast-item-img"
                draggable="false"
                src="${item.weatherImgSrc}"
                loading="lazy"
                height="45"
                width="45"
              />
              <div>
                <span part="forcast-item-high">${item.high}°</span>
                <span part="forcast-item-low">${item.low}°</span>
              </div>
            </div>
          `,
        )}
      </div>
    `;
  }

  // eslint-disable-next-line class-methods-use-this
  get loadingTm() {
    return html`
      <div part="loading">
        <div>• • •</div>
      </div>
    `;
  }

  get errorTm() {
    return html`
      <div class="horizontal center-center">
        <img
          part="error-img"
          draggable="false"
          src="${this.iconBaseUrl}/unknown.png"
          loading="lazy"
          height="120"
          width="120"
        />
      </div>
      <div part="error-text">Error getting weather</div>
    `;
  }

  static get properties() {
    return {
      // openweathermap API key
      apiKey: { type: String },
      // See https://openweathermap.org/current
      // {city name},
      // {city name}, {state}
      // {city name}, {state}, {country code}
      // {zip code}
      // {zip code},{country code}
      query: { type: String },
      // True when data is loading
      loading: { type: Boolean },
      // 'Metric' or 'Imperial'
      units: { type: String },
      // Current day of the week
      weekDayName: { type: String },
      // Time when data was last requested
      time: { type: String },
      // Set to true to use open weather icons
      useOpenWeatherIcons: { type: Boolean },
      // BaseUrl for custom icons
      // defaults to '/icons'
      iconBaseUrl: { type: String },
      // Data object
      data: { type: Object },
      // Forecast object
      forecast: { type: Array },
      error: { type: Boolean },
    };
  }

  constructor() {
    super();
    this.query = '';
    this.data = {};
    this.loading = true;
    this.units = 'Imperial';
    // Set date and times
    const today = new Date();
    this.weekDayName = getWeekDay(today.getDay());
    this.time = getLocalTime(today);
    this.useOpenWeatherIcons = false;
    this.forecast = [];
    this.iconBaseUrl = '/icons';
    this.error = false;
  }

  updated(changed) {
    if (changed.has('apiKey') || changed.has('query') || changed.has('units')) {
      this._generateRequests(this.apiKey, this.query);
    }
  }

  toggleUnits() {
    this.units = this.units === 'Imperial' ? 'Metric' : 'Imperial';
  }

  refresh() {
    this.time = getLocalTime(new Date());
    this._generateRequests(this.apiKey, this.query);
  }

  _generateRequests(key, query) {
    if (!key || !query) return;
    this.error = false;
    this.loading = true;

    const locationParam = this._getLocationParam(query);
    this._getCurrentWeather(key, locationParam);
    this._getForecast(key, locationParam);
  }

  // eslint-disable-next-line class-methods-use-this
  _getLocationParam(query) {
    // eslint-disable-next-line no-restricted-globals
    const firstChar = query.toString().charAt(0);
    const isZipCode = !isNaN(firstChar - parseFloat(firstChar));
    return isZipCode ? `zip=${query}` : `q=${query}`;
  }

  _getCurrentWeather(key, locationParam) {
    const url = `https://api.openweathermap.org/data/2.5/weather?${locationParam}&appid=${key}&units=${this.units}`;
    fetch(url)
      .then(response => {
        if (!response.ok) throw new Error('Something went wrong');
        return response.json();
      })
      .then(data => {
        this.loading = false;
        this.data = {
          id: data.id,
          name: data.name,
          sunrise: getLocalTime(new Date(data.sys.sunrise * 1000)),
          sunset: getLocalTime(new Date(data.sys.sunset * 1000)),
          temp: +data.main.temp.toFixed(0),
          description: data.weather[0].description,
          humidity: data.main.humidity,
          wind: `${data.wind.speed.toFixed(0)} ${this.units === 'Imperial' ? 'mph' : 'km/h'}`,
          weatherImgSrc: this._computeWeatherImgSrc({
            icon: data.weather[0].icon,
            id: data.weather[0].id,
            dynamicDayOrNightString: true,
          }),
          fullForecastUrl: `https://openweathermap.org/city/${data.id}`,
        };
        this.dispatchEvent(new CustomEvent('data-changed', { detail: { value: this.data } }));
      })
      .catch(error => {
        this._handleWeatherError(error);
      });
  }

  _getForecast(key, locationParam) {
    const url = `https://api.openweathermap.org/data/2.5/forecast?${locationParam}&appid=${key}&units=${this.units}`;
    fetch(url)
      .then(response => {
        if (!response.ok) throw new Error('Something went wrong');
        return response.json();
      })
      .then(data => {
        this.loading = false;
        this.forecast = this._computeForecast(data.list);
      })
      .catch(error => {
        this._handleWeatherError(error);
      });
  }

  // Handle fetch error's
  _handleWeatherError(err) {
    this.loading = false;
    this.error = true;
    this.data = {};
    this.forecast = [];
    // eslint-disable-next-line no-console
    console.error('Error getting lit-weather data', err);
  }

  // Compute the forecast as best as we can with the data returned from openweathermap
  // This computation is being done because the full day forecast requires a paid account
  // PR's welcome here to refactor how this is done, there is room for improvement
  _computeForecast(items) {
    if (!items || !items.length) return [];

    let data = {};
    for (const item of items) {
      // item forecast date object
      const forecastDate = new Date(item.dt_txt);
      // item unique forecast date key
      const forecastDateKey = `${forecastDate.getMonth() + 1}-${forecastDate.getDate()}`;

      // Existing "Day" data
      const day = data[forecastDateKey] || {};

      // Highs and lows
      // Current hours high
      const high = item.main.temp_max;
      // Current hours low
      const low = item.main.temp_min;
      // Set new high if this hours high is greater than what aready exists or if no existing value is present
      day.high = !day.high || day.high < high ? +high.toFixed(0) : day.high;
      // Set new low if this hours low is less than what aready exists or if no existing value is present
      day.low = !day.low || day.low > low ? +low.toFixed(0) : day.low;

      // Set week day label
      if (!day.weekDayName) day.weekDayName = getWeekDay(forecastDate.getDay());

      // Weather codes
      const weather = item.weather[0];
      if (!day.weatherCodes) day.weatherCodes = [];
      day.weatherCodes = [[weather.icon, weather.id], ...day.weatherCodes];

      // Set day data in the object array
      data[forecastDateKey] = day;
    }

    // Attempt to look for the most prominent weather codes for each day
    // https://stackoverflow.com/a/20762713
    // prettier-ignore
    const getMostProminentWeatherId = (arr) => arr.sort((a, b) => arr.filter(v => v === a).length - arr.filter(v => v === b).length).pop();

    // Convert data object to array and map weatherImgSrc accordingly
    data = Object.keys(data).map(k => {
      const codes = data[k].weatherCodes;
      const icon = getMostProminentWeatherId(codes.map(code => code[0]));
      const id = getMostProminentWeatherId(codes.map(code => code[1]));
      return {
        weatherImgSrc: this._computeWeatherImgSrc({ icon, id }),
        ...data[k],
      };
    });
    // Remove the last item, it was very inaccurate and only contained one record from my testing
    data.pop();

    return data;
  }

  _computeWeatherImgSrc({ icon, id, dynamicDayOrNightString }) {
    if (!icon && !id) return '';
    if (this.useOpenWeatherIcons && icon) return `http://openweathermap.org/img/wn/${icon}@2x.png`;

    let dayOrNightStr = 'day';
    // When setting the current weather img, show the icon for day or night according to the icon string returned from openWeather
    // This is so we do not have to compare any dates in the local timezone
    if (dynamicDayOrNightString) {
      dayOrNightStr = icon.indexOf('d') > -1 ? 'day' : 'night';
    }

    // https://openweathermap.org/weather-conditions
    switch (id) {
      case 200:
      case 201:
      case 202:
      case 210:
      case 211:
      case 212:
      case 221:
      case 230:
      case 231:
      case 232:
        return `${this.iconBaseUrl}/thunder-${dayOrNightStr}.png`;
      case 300:
      case 301:
      case 302:
      case 303:
      case 304:
      case 311:
      case 320:
      case 321:
      case 322:
      case 331:
      case 500:
      case 501:
      case 502:
      case 503:
      case 504:
      case 511:
      case 520:
      case 521:
      case 522:
      case 531:
        return `${this.iconBaseUrl}/rainy-${dayOrNightStr}.png`;
      case 600:
      case 601:
      case 602:
      case 611:
      case 612:
      case 613:
      case 615:
      case 616:
      case 620:
      case 621:
      case 622:
        return `${this.iconBaseUrl}/snow-${dayOrNightStr}.png`;
      case 721:
      case 741:
      case 751:
      case 761:
      case 762:
      case 771:
        return `${this.iconBaseUrl}/haze-${dayOrNightStr}.png`;
      case 731:
      case 781:
        return `${this.iconBaseUrl}/windy-${dayOrNightStr}.png`;
      case 800:
        return `${this.iconBaseUrl}/clear-${dayOrNightStr}.png`;
      case 801:
        return `${this.iconBaseUrl}/partly-cloudy-${dayOrNightStr}.png`;
      case 802:
      case 803:
        return `${this.iconBaseUrl}/mostly-cloudy-${dayOrNightStr}.png`;
      case 804:
        return `${this.iconBaseUrl}/cloudy-weather.png`;
      default:
        if (icon) return `http://openweathermap.org/img/wn/${icon}@2x.png`;
        return `${this.iconBaseUrl}/unknown.png`;
    }
  }
}


demo/demo-element.js
import { LitElement, html, css } from 'lit';
import '../lit-weather';
import { unsafeHTML } from 'lit/directives/unsafe-html.js';

const defaultTabClassName = /*tw*/ 'rounded-md px-3 py-2 text-sm font-medium';
const activeTabClassName =
  /*tw*/ 'bg-gray-50/50 dark:bg-white/10 dark:text-white text-black rounded-md px-3 py-2 text-sm font-medium';

class DemoElement extends LitElement {
  render() {
    return html`
      <style>
        :host {
          display: block;
        }

        a {
          text-decoration: underline;
        }
      </style>

      <div class=${`${this.variant === 'horizontal' ? 'max-w-2xl w-full' : 'max-w-xl'} mx-auto`}>
        <div class="gap-x-4 mb-10">
          <h2 class="text-3xl font-bold tracking-tight sm:text-4xl">lit-weather</h2>
          <p class="mx-auto mt-6 text-lg leading-8 dark:text-gray-300 text-gray-600">
            <a target="_blank" href="https://lit.dev" class="text-underline">Lit</a> powered weather
            forecast
            <a href="https://developer.mozilla.org/en-US/docs/Web/Web_Components">web component</a>,
            leveraging the <a href="https://openweathermap.org/">OpenWeatherMap</a> API. It's
            customizable with <a href="https://tailwindcss.com/">tailwindcss</a>,
            <a href="https://bundlephobia.com/package/lit-weather">lightweight</a>, and shadow dom
            free.
          </p>
        </div>

        <nav
          class="inline-flex items-center justify-center bg-white rounded-md bg-gray-200/50 dark:bg-white/5 p-1 mb-10 text-gray-700 dark:text-gray-300"
        >
          <button
            id="santa-barbara"
            @click="${() => {
              this.value = 'Santa Barbara';
            }}"
            class="${this.value === 'Santa Barbara' ? activeTabClassName : defaultTabClassName}"
          >
            Santa Barbara
          </button>
          <button
            id="sydney"
            @click="${() => {
              this.value = 'Sydney, Australia';
            }}"
            class="${this.value === 'Sydney, Australia' ? activeTabClassName : defaultTabClassName}"
          >
            Sydney, Australia
          </button>
          <button
            id="90210"
            @click="${() => {
              this.value = '90210';
            }}"
            class="${this.value === '90210' ? activeTabClassName : defaultTabClassName}"
          >
            90210
          </button>
          <button
            id="california"
            @click="${() => {
              this.value = 'California';
            }}"
            class="${this.value === 'California' ? activeTabClassName : defaultTabClassName}"
          >
            California
          </button>
        </nav>

        <lit-weather
          class="block"
          .query="${this.value}"
          .variant="${this.variant}"
          token="1e366c8dec8ecdfd589dd13d8aa454e2"
        ></lit-weather>

        <nav
          class="inline-flex mt-10 items-center justify-center bg-white rounded-md bg-gray-200/50 dark:bg-white/5 p-1 mb-10 text-gray-700 dark:text-gray-300"
        >
          <button
            id="stacked"
            @click="${() => {
              this.variant = 'stacked';
            }}"
            class="${!this.variant || this.variant === 'stacked'
              ? activeTabClassName
              : defaultTabClassName}"
          >
            Stacked
          </button>
          <button
            id="preview"
            @click="${() => {
              this.variant = 'preview';
            }}"
            class="${this.variant === 'preview' ? activeTabClassName : defaultTabClassName}"
          >
            Preview
          </button>
        </nav>

        <div class="mt-16">
          <h2 class="text-3xl mb-5">Properties</h2>
          <div class="my-6">
            <ul
              role="list"
              class="m-0 max-w-[calc(theme(maxWidth.lg)-theme(spacing.8))] list-none divide-y divide-zinc-900/5 p-0 dark:divide-white/5"
            >
              ${this.properties && this.properties.length > 0
                ? this.properties.map(
                    p => html`
                      <div class="m-0 px-0 py-4 first:pt-0 last:pb-0">
                        <dl class="m-0 flex flex-wrap items-center gap-x-3 gap-y-2">
                          <dt class="sr-only">Name</dt>
                          <dd>
                            <code>${p.name}</code>
                          </dd>

                          ${p.type.text
                            ? html`
                                <dt class="sr-only">Type</dt>
                                <dd class="font-mono text-xs text-zinc-400 dark:text-zinc-500">
                                  ${p.type.text}
                                  ${['query', 'token'].includes(p.name) ? ' | required' : ''}
                                </dd>
                              `
                            : ''}

                          <dt class="sr-only">Description</dt>
                          <dd
                            class="w-full !text-sm text-left flex-none text-gray-500 [&gt;:first-child]:mt-0 [&gt;:last-child]:mb-0"
                          >
                            ${unsafeHTML(p.description)}
                          </dd>
                        </dl>

                        ${p.name !== 'cn'
                          ? html`${p.default
                              ? html`
                                  <code class="whitespace-pre text-[10px] mt-4 block text-gray-500"
                                    >Default: ${p.default}</code
                                  >
                                `
                              : ''}`
                          : html`<button
                                id="${p.name}-button"
                                class="text-[10px] mt-3 text-gray-400 group"
                                @click="${() => {
                                  document.getElementById(p.name).classList.toggle('hidden');
                                  document
                                    .getElementById(p.name + '-button')
                                    .toggleAttribute('open');
                                }}"
                              >
                                <span class="group-open:hidden">View Default</span>
                                <span class="hidden group-open:block">Hide Default</span>
                              </button>
                              <div class="hidden" id="${p.name}">
                                ${p.default
                                  ? html`
                                      <code
                                        class="whitespace-pre text-[10px] mt-4 block text-gray-500"
                                        >${p.default}</code
                                      >
                                    `
                                  : ''}
                              </div>`}
                        ${p.typeDef
                          ? html`
                              <code class="whitespace-pre text-[10px] mt-4 block text-gray-500"
                                >${p.typeDef}</code
                              >
                            `
                          : ''}
                      </div>
                    `
                  )
                : html`<p>Loading...</p>`}
            </ul>
          </div>
        </div>

        <div class="mt-16 block text-xs text-gray-500">
          <a target="_blank" href="https://codepen.io/ryanburns23/pen/vYzgVLE">
            Edit on CodePen &rarr;
          </a>
        </div>
      </div>
    `;
  }

  static get properties() {
    return {
      value: { type: String },
      variant: { type: String },

      properties: { type: Object },
    };
  }

  constructor() {
    super();
    this.value = 'Santa Barbara';
    fetch('/custom-elements.json').then(response => {
      response.json().then(json => {
        this.properties = json.modules[0].declarations[0].members.filter(
          m => m.kind === 'field' && m.type && m.privacy !== 'protected'
        );
      });
    });
  }

  refresh() {
    this.requestUpdate();
    console.log(this.value);
    document.querySelector('lit-weather').query = this.value;
    document.querySelector('lit-weather').refresh();
  }

  createRenderRoot() {
    return this;
  }
}

customElements.define('demo-element', DemoElement);
